qiskit/circuit/quantumcircuit.py

```diff
@@ -285,6 +286,65 @@ def __init__(
             raise TypeError("Only a dictionary or None is accepted for circuit metadata")
         self._metadata = metadata
 
+    @staticmethod
+    def from_instructions(
+        instructions: Iterable[CircuitInstruction],
+        *,
+        name: Optional[str] = None,
+        global_phase: ParameterValueType = 0,
+        metadata: Optional[dict] = None,
+    ) -> "QuantumCircuit":
+        """Construct a circuit from an iterable of CircuitInstructions.
+
+        Args:
+            instructions: The instructions to add to the circuit.
+            name: The name of the circuit.
+            global_phase: The global phase of the circuit in radians.
+            metadata: Arbitrary key value metadata to associate with the circuit.
+
+        Returns:
+            The quantum circuit.
+        """
+        circuit = QuantumCircuit(name=name, global_phase=global_phase, metadata=metadata)
+        added_qubits = set()
+        added_clbits = set()
+        for instruction in instructions:
+            _, qubits, clbits = instruction
+            qubits = [qubit for qubit in qubits if qubit not in added_qubits]
+            clbits = [clbit for clbit in clbits if clbit not in added_clbits]
+            circuit.add_bits(qubits)
+            circuit.add_bits(clbits)
+            added_qubits.update(qubits)
+            added_clbits.update(clbits)
+            circuit._append(instruction)
+        return circuit
+
+    @staticmethod
+    def from_instruction_tuples(
+        instructions: Iterable[tuple[Instruction, Iterable[Qubit], Iterable[Clbit]]],
```

Comment:
Do we want two methods for this or rather handle both types in one? Handling both in one would have the advantage that you could use `from_instructions` with both the old and new format of circuit data, plus I guess we want to promote using `CircuitInstruction` over the tuples? 😄

Reply:
I combined both methods. The main reason I want to support the tuples format is to reduce boilerplate in my own code. For example, in this code below, the tuples syntax is more concise https://github.com/Qiskit/qiskit-terra/blob/df40d765c902279f7efc9e8ad17b04094429bb3e/test/python/circuit/test_circuit_operations.py#L1281-L1293